addons/l10n_id_efaktur_coretax/models/efaktur_document.py

```diff
@@ -0,0 +1,86 @@
+    )
+    invoice_ids = fields.One2many(
+        comodel_name="account.move",
+        inverse_name="l10n_id_efaktur_document",
+        domain="[('move_type', 'in', ['out_invoice', 'out_refund']), ('company_id', '=', company_id), ('l10n_id_efaktur_document', '=', False), ('state', '=', 'posted')]",
+    )
+    attachment_id = fields.Many2one(comodel_name="ir.attachment", readonly=True)
+
+    def _compute_name(self):
+        for doc in self:
+            sorted_invoices = doc.invoice_ids.sorted('name')
+            name = []
+            if sorted_invoices:
+                name.append(sorted_invoices[0].name)
+                if len(sorted_invoices) > 1:
+                    name.append(sorted_invoices[-1].name)
+            doc.name = "%s - Efaktur (%s)" % (fields.Date.context_today(doc).strftime("%Y%m%d"), "....".join(name))
+
+    def action_download(self):
+        """ Download E-Faktur of related attachment """
+        for document in self:
+            if not document.attachment_id:
+                document._generate_xml()
+        return {
+            'type': 'ir.actions.act_url',
+            'url': f'/l10n_id_efaktur_coretax/download_attachments/{",".join(map(str, self.attachment_id.ids))}',
+        }
+
+    def _generate_xml(self):
+        """ Generate the XML file as content and save it as attachment in this record"""
+        self.ensure_one()
+
+        if self.invoice_ids.filtered(lambda x: not x.l10n_id_kode_transaksi):
+            raise UserError(_("Some documents don't have a transaction code"))
+        if self.invoice_ids.filtered(lambda x: x.move_type != 'out_invoice'):
+            raise UserError(_("Some documents are not Customer Invoices"))
+        raw_data = self._generate_efaktur_invoice()
+
+        if not self.attachment_id:
+            attachment = self.env['ir.attachment'].create({
+                'raw': raw_data,
+                'name': 'efaktur_%s.xml' % (fields.Datetime.to_string(fields.Datetime.now()).replace(" ", "_")),
+                'type': 'binary',
+                'res_model': 'l10n_id_efaktur_coretax.document',
+                'res_id': self.id,
+            })
+            self.attachment_id = attachment.id
+        else:
+            attachment = self.attachment_id
+            self.attachment_id.write({
+                'raw': raw_data,
+                'name': 'efaktur_%s.xml' % (fields.Datetime.to_string(fields.Datetime.now()).replace(" ", "_")),
+            })
+
+        self.message_post(
+            body=_("The e-Faktur report has been generated"),
+            attachments=[(attachment.name, attachment.raw)]
+        )
+
+    def _generate_efaktur_invoice(self):
+        """ Generate E-Faktur for customer invoice. Prepare data, load XML template and """
+        invoice_data = self.invoice_ids.prepare_efaktur_vals()
+        xml_content = self.env['ir.qweb']._render('l10n_id_efaktur_coretax.efaktur_coretax_template', {'data': invoice_data, 'TIN': self.company_id.vat})
```

Comment:
Why not use the TIN in the data :thinking:

Reply:
Referring to https://docs.google.com/spreadsheets/d/1bHBCjYSag1weyaL_q_-Zrt9sM9tW-Sw-blnsdkEcJaw/edit?gid=0#gid=0, 

they're expecting TIN to be put before declaring list of invoices, so I'm putting it outside the invoice vals